### PR TITLE
Add restriction for unique test names

### LINKED ISSUE #1387: Suggestion: enforce unique GoogleTest names
Description

Many tests in Stan Math have duplicate test names. You can check this with the appropriate regex | sort | uniq pipe, (e.g. rg -INo "TEST(_F|_P)?\(\w*, \w*\)" | sort | uniq -d)

Some tests also have the same name that are part of a test fixture, and some that are not. This is an error in GTest.

Expected:

All unit tests should have unique names.

Reason:

The GTest test runner supports filtering and running based on test name. This is also true for the CMake CTest runner. In general, when you run all the unit tests, having unique test names is necessary once you move beyond compiling individual test executables through runTests.py in order to easily see what tests are passing and failing. Also, it's a compilation error for two GTest tests to have the same name.

Eventually, I would like to support jumbo (unity) builds of the unit tests. Having unique test names is a prerequisite.

I used to maintain my own patches to the sources here, but it was an automated thing and the names were not chosen well, just the existing ones with 1 or 2 appended to the test, so I have discarded them. I would like to post this issue here just to let people know that if they make new tests, to please make sure the names are unique.

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

Fixes #1387 

This PR adds the restriction of having unique names for `TEST()` and `TEST_F()` across the entire test/unit folder. It does so by adding a check to `runChecks.py` that looks for duplicates in all `test/unit/*_test.cpp` files.

Developers can verify they have not introduced non-unique names by running `python runChecks.py`.

The first commit will be the python script only. That way we can see the errors on Jenkins to verify that it works OK there. Then I will merge the cleanup branch that fixes all the errors.

## Tests

/

## Side Effects

/

## Checklist

- [x] Math issue #1387

- [x] Copyright holder: Rok Češnovar

- [x] the basic tests are passing

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
You can check how the error look like on Jenkins https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-1422/1/pipeline/

An example: 
```
Tests or test fixtures with non-unqiue names found in test/unit:

(MathMatrix, matrix_exp_pade_1x1) in files:
	test/unit/math/rev/mat/fun/matrix_exp_pade_test.cpp
	test/unit/math/fwd/mat/fun/matrix_exp_pade_test.cpp
	test/unit/math/prim/mat/fun/matrix_exp_pade_test.cpp
(MathMatrix, matrix_exp_multiply_exception) in files:
	test/unit/math/rev/mat/fun/matrix_exp_multiply_test.cpp
	test/unit/math/prim/mat/fun/matrix_exp_multiply_test.cpp
(ErrorHandlingScalar, CheckNonnegative_nan) in files:
	test/unit/math/prim/scal/err/check_nonnegative_test.cpp
	test/unit/math/prim/arr/err/check_nonnegative_test.cpp
(MathMeta, value_type) in files:
	test/unit/math/rev/scal/meta/child_type_test.cpp
	test/unit/math/fwd/scal/meta/child_type_test.cpp
	test/unit/math/prim/scal/meta/child_type_test.cpp
	test/unit/math/prim/arr/meta/value_type_test.cpp
	test/unit/math/mix/scal/meta/child_type_test.cpp
```

**Comment by rok-cesnovar** [comment]
This one is going to be a pain to review because this touches A LOT of files
So to help out here are the changes that are not just renaming tests:
- 39395b2d removes unused tests that were behind `#if 0`
- 8ecff0ae deletes a test file that was never used (fma.hpp cant be triggered) and adds non-duplicates tests to fma_test.cpp
- c80b13ea removes duplicate tests 

The rest just rename tests. I tried to make some of the generic labels like MathMatrix more specific with at least MathMatrixRevScal or something along those lines.

**Comment by rok-cesnovar** [comment]
Also this will probably cause merge issues with @bob-carpenter's efforts on the test framework. I am also happy to hold of on this PR until that gets merged.

**Comment by alashworth** [comment]
This is great, thanks for adding this to the main branch.

File: test/unit/math/rev/mat/functor/integrate_ode_adams_rev_test.cpp [hunk 1]
```diff
@@ -98,8 +98,7 @@ void sho_error_test(F harm_osc, std::vector<double>& y0, double t0,
                    std::invalid_argument, error_msg);
 }
 
-// TODO(carpenter): g++6 failure
-TEST(StanAgradRevOde_integrate_ode, harmonic_oscillator_finite_diff) {
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply test/unit/math/rev/mat/functor/integrate_ode_adams_rev_test.cpp]: I tried this one a AWS instance with g++6 and found no errors and I am assuming this one is done.

File: test/unit/math/rev/mat/functor/integrate_ode_bdf_rev_test.cpp [hunk 2]
```diff
@@ -98,8 +98,7 @@ void sho_error_test(F harm_osc, std::vector<double>& y0, double t0,
       std::invalid_argument, error_msg);
 }
 
-// TODO(carpenter): g++6 failure
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply test/unit/math/rev/mat/functor/integrate_ode_bdf_rev_test.cpp]: Same as above I ran all these tests on g++6 and found nothing wrong. If there is something wrong we should open an issue.

File: test/unit/math/rev/mat/functor/integrate_ode_bdf_rev_test.cpp [hunk 3]
```diff
@@ -137,8 +136,7 @@ TEST(StanAgradRevOde_integrate_ode, harmonic_oscillator_error) {
   sho_error_test<var, var>(harm_osc, y0, t0, ts, theta, x, x_int, error_msg);
 }
 
-// TODO(carpenter): g++6 failure
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply test/unit/math/rev/mat/functor/integrate_ode_bdf_rev_test.cpp]: Same as above.

File: test/unit/math/prim/mat/fun/cov_matrix_transform_test.cpp [hunk 4]
```diff
@@ -7,35 +7,6 @@ using Eigen::Matrix;
...
-  y.resize(10, 0);
-  EXPECT_THROW(stan::math::cov_matrix_free_lkj(y), std::invalid_argument);
-  y.resize(1, 2);
-  EXPECT_THROW(stan::math::cov_matrix_free_lkj(y), std::invalid_argument);
-
-  y.resize(2, 2);
-  y << 0, 0, 0, 0;
-  EXPECT_THROW(stan::math::cov_matrix_free_lkj(y), std::domain_error);
-}
-
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply test/unit/math/prim/mat/fun/cov_matrix_transform_test.cpp]: Here https://github.com/stan-dev/math/blob/develop/test/unit/math/prim/mat/fun/lkj_cov_matrix_transform_test.cpp

**Comment by rok-cesnovar** [comment]
Thanks for the python script comments! I think I addressed everything.

**Comment by rok-cesnovar** [comment]
There were some additional non-unique tests that were found after the merge. The push dismissed the review :/

**Comment by syclik** [comment]
@rok-cesnovar, somehow the tests haven't passed. Can you kick this off again?

Also, with this PR, has this been integrated into the continuous integration?

**Comment by rok-cesnovar** [comment]
Yeah, some upstream tests timeouted or something weird. Restarted now. 

These checks are part of runChecks.py that is run with make test-math-dependencies (https://github.com/stan-dev/math/blob/develop/Jenkinsfile#L131). Those are run in the first phases of a Jenkins tests.

An example of what the check does on develop (develop when I started this PR to be exact) is given here: 
https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-1422/1/pipeline/

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal  sorry to bother you again, this seems like a weird failure. It might be some issue I caused on this PR, but I am seeing no errors printed: https://jenkins.mc-stan.org/blue/organizations/jenkins/Math%20Pipeline/detail/PR-1422/33/pipeline

Thanks!

**Comment by serban-nicusor-toptal** [comment]
Fixed, was trying to run on the new windows machine.

**Pull request merged**